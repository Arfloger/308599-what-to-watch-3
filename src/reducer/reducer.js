import CommentModel from "../models/comment-model";
const MIN_CARDS_ON_PAGE = 8;

export const initialState = {
  movies: [],
  promo: {},
  comments: [],
  moviesFavorite: [],
  currentMovie: JSON.parse(window.localStorage.getItem(`currentMovie`)),
  isAuthorized: true,
  avatar: undefined,
  cardsOnPage: MIN_CARDS_ON_PAGE,
};

const cashCurrentMovie = (movie) => {
  window.localStorage.setItem(`currentMovie`, JSON.stringify(movie));

  return movie;
};

export const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
  LOAD_PROMO: `LOAD_PROMO`,
  LOAD_COMMENTS: `LOAD_COMMENTS`,
  LOAD_FAVORITE_MOVIES: `LOAD_FAVORITE_MOVIES`,
  CHECK_AUTHORIZATION: `CHECK_AUTHORIZATION`,
  GET_AVATAR: `GET_AVATAR`,
  UPDATE_CURRENT_MOVIE: `UPDATE_CURRENT_MOVIE`,
  INCREASE_QUANTITY_FILMS: `INCREASE_QUANTITY_FILMS`,
  RESET_TO_MIN_FILMS: `RESET_TO_MIN_FILMS`,
};

export const ActionCreator = {
  loadMovies: (movies) => ({
    type: ActionType.LOAD_MOVIES,
    payload: movies
  }),
  loadPromo: (promo) => ({
    type: ActionType.LOAD_PROMO,
    payload: promo
  }),
  loadComments: (comments) => ({
    type: ActionType.LOAD_COMMENTS,
    payload: comments
  }),
  loadFavoriteMovies: (movies) => ({
    type: ActionType.LOAD_FAVORITE_MOVIES,
    payload: movies
  }),
  checkAuthorization: (isAuth) => ({
    type: ActionType.CHECK_AUTHORIZATION,
    payload: isAuth
  }),
  getAvatar: (avatarUrl) => ({
    type: ActionType.GET_AVATAR,
    payload: avatarUrl
  }),
  updateCurrentMovie: (movieId) => ({
    type: ActionType.UPDATE_CURRENT_MOVIE,
    payload: movieId
  }),
  increaseQuantityFilms: () => ({
    type: ActionType.INCREASE_QUANTITY_FILMS,
    payload: 8,
  }),
  resetToMinFilms: () => ({
    type: ActionType.RESET_TO_MIN_FILMS,
    payload: MIN_CARDS_ON_PAGE,
  }),
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_MOVIES:
      return Object.assign({}, state, {
        movies: action.payload
      });
    case ActionType.LOAD_PROMO:
      return Object.assign({}, state, {
        promo: action.payload
      });
    case ActionType.LOAD_COMMENTS:
      return Object.assign({}, state, {
        comments: CommentModel.parseComments(action.payload)
      });
    case ActionType.LOAD_FAVORITE_MOVIES:
      return Object.assign({}, state, {
        moviesFavorite: action.payload
      });
    case ActionType.CHECK_AUTHORIZATION:
      return Object.assign({}, state, {
        isAuthorized: action.payload
      });
    case ActionType.GET_AVATAR:
      return Object.assign({}, state, {
        avatar: action.payload
      });
    case ActionType.UPDATE_CURRENT_MOVIE:
      return Object.assign({}, state, {
        currentMovie: cashCurrentMovie(state.movies[action.payload])
      });
    case ActionType.INCREASE_QUANTITY_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: state.cardsOnPage + action.payload,
      });
    case ActionType.RESET_TO_MIN_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: action.payload,
      });
  }
  return state;
};

export const Operation = {
  loadMovies: () => (dispatch, _getState, api) => {
    return api.get(`/films`)
      .then(({data}) => {
        dispatch(ActionCreator.loadMovies(data));
      });
  },
  loadPromo: () => (dispatch, _getState, api) => {
    return api.get(`/films/promo`)
      .then(({data}) => {
        dispatch(ActionCreator.loadPromo(data));
      });
  },
  loadComments: (movieId) => (dispatch, _getState, api) => {
    return api.get(`/comments/${movieId}`)
      .then(({data}) => {
        dispatch(ActionCreator.loadComments(data));
      });
  },
  loadFavoriteMovies: () => (dispatch, _getState, api) => {
    return api.get(`/favorite`)
      .then(({data}) => {
        dispatch(ActionCreator.loadFavoriteMovies(data));
      });
  },
  addReview: (movieId, body, cb) => (dispatch, _getState, api) => {
    let error = false;
    return api.post(`/comments/${movieId}`, body)
      .then(({data}) => {
        error = false;
        cb(error);
        dispatch(ActionCreator.loadComments(data));
      })
      .catch(() => {
        error = true;
        cb(error);
      });
  },
  switchFavorite: (movieId, status, cb) => (dispatch, _getState, api) => {
    return api.post(`/favorite/${movieId}/${status}`)
      .then(() => {
        cb();
      });
  },
  logIn: (body) => (dispatch, _getState, api) => {
    return api.post(`/login`, body)
      .then(({data}) => {
        dispatch(ActionCreator.checkAuthorization(true));
        dispatch(ActionCreator.getAvatar(data.avatar_url));
      })
      .catch(() => {
        dispatch(ActionCreator.checkAuthorization(false));
      });
  },
  checkAuthorization: () => (dispatch, _getState, api) => {
    return api.get(`/login`)
      .then(({data}) => {
        dispatch(ActionCreator.checkAuthorization(true));
        dispatch(ActionCreator.getAvatar(data.avatar_url));
      })
      .catch(() => {
        dispatch(ActionCreator.checkAuthorization(false));
      });
  }
};
