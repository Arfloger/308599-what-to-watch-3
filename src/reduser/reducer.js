export const getGenreList = (films) => {

  let genresList = new Set(films.map((it) => it[`genre`]));
  initialState.genres = initialState.genres.concat(Array.from(genresList)).slice(0, 10);

  return initialState.genres;
};

export const filteredFilms = (genre) => {
  if (genre === `All genres`) {
    return initialState.films.slice();
  }

  return initialState.films.slice().filter((it) => it.genre === genre);
};

export const getFilmById = (id) => {
  initialState.currentFilmCard = initialState.films.slice().find((item) => item.id == id);

  return initialState.currentFilmCard;
};

const MIN_CARDS_ON_PAGE = 8;

export const initialState = {
  genre: `All genres`,
  genres: [`All genres`],
  films: [],
  cardsOnPage: MIN_CARDS_ON_PAGE,
  requireAuthorization: false,
  avatar: null,
  currentFilmCard: null,
  promo: null,
};

const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  LOAD_FILMS: `LOAD_FILMS`,
  LOAD_PROMO: `LOAD_PROMO`,
  FILTERED_FILMS: `FILTERED_FILMS`,
  GET_GENRE_LIST: `GET_GENRE_LIST`,
  INCREASE_QUANTITY_FILMS: `INCREASE_QUANTITY_FILMS`,
  RESET_TO_MIN_FILMS: `RESET_TO_MIN_FILMS`,
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  LOGIN_USER: `LOGIN_USER`,
  GET_FILM_BY_ID: `GET_FILM_BY_ID`,
};

export const ActionCreator = {
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    payload: genre,
  }),

  loadFilms: (films) => ({
    type: ActionType.LOAD_FILMS,
    payload: films,
  }),

  loadPromo: (promo) => ({
    type: ActionType.LOAD_PROMO,
    payload: promo,
  }),

  filteredFilms: (genre) => ({
    type: ActionType.FILTERED_FILMS,
    payload: filteredFilms(genre),
  }),

  getGenreList: (genres) => ({
    type: ActionType.GET_GENRE_LIST,
    payload: getGenreList(genres),
  }),

  increaseQuantityFilms: () => ({
    type: ActionType.INCREASE_QUANTITY_FILMS,
    payload: 20,
  }),

  resetToMinFilms: () => ({
    type: ActionType.RESET_TO_MIN_FILMS,
    payload: MIN_CARDS_ON_PAGE,
  }),

  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status,
  }),

  loginUser: (data) => ({
    type: ActionType.LOGIN_USER,
    payload: data,
  }),
  getFilmById: (id) => ({
    type: ActionType.GET_FILM_BY_ID,
    payload: getFilmById(id),
  }),
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        genre: action.payload,
      });
    case ActionType.LOAD_FILMS:
      return Object.assign({}, state, {
        films: action.payload,
      });
    case ActionType.LOAD_PROMO:
      return Object.assign({}, state, {
        promo: action.payload,
      });
    case ActionType.GET_GENRE_LIST:
      return Object.assign({}, state, {
        genres: action.payload,
      });
    case ActionType.FILTERED_FILMS:
      return Object.assign({}, state, {
        films: action.payload,
      });
    case ActionType.INCREASE_QUANTITY_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: state.cardsOnPage + action.payload,
      });
    case ActionType.RESET_TO_MIN_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: action.payload,
      });
    case ActionType.REQUIRE_AUTHORIZATION:
      return Object.assign({}, state, {
        requireAuthorization: action.payload,
      });
    case ActionType.LOGIN_USER:
      return Object.assign({}, state, {
        id: action.payload.id,
        email: action.payload.email,
        name: action.payload.name,
        avatar: action.payload.avatar_url,
      });
    case ActionType.GET_FILM_BY_ID:
      return Object.assign({}, state, {
        currentFilmCard: action.payload,
      });
  }

  return state;
};

export const Operation = {
  loadFilms: () => (dispatch, _getState, api) => {
    return api.get(`/films`)
      .then((response) => {
        dispatch(ActionCreator.loadFilms(response.data));
        dispatch(ActionCreator.getGenreList(response.data));
        initialState.films = response.data;
      })
      .catch((error) => {
        throw new Error(error);
      });
  },
  loadPromo: () => (dispatch, _getState, api) => {
    return api.get(`/films/promo`)
      .then((response) => {
        dispatch(ActionCreator.loadPromo(response.data));
        initialState.promo = response.data;
      });
  },
  checkAuth: (login, password) => {
    return (dispatch, _getState, api) => {
      return api
        .post(`/login`, {
          email: login,
          password,
        })
        .then((res) => {
          if (res.status >= 200) {
            dispatch(ActionCreator.requireAuthorization(true));
            dispatch(ActionCreator.loginUser(res.data));
          }
        })
        .catch((error) => {
          throw new Error(error);
        });
    };
  }
};
