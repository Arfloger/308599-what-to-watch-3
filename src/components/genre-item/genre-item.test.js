import React from "react";
import renderer from "react-test-renderer";
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {reducer} from "../../reducer";

import GenreItem from "./genre-item.jsx";

it(`GenreItem correctly renders after relaunch`, () => {
  const store = createStore(reducer);
  const tree = renderer
    .create(<Provider store={store}><GenreItem
      genreName={``}
      activeTab={false}
      onGenreTabClick={() => {}}
    /></Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});
