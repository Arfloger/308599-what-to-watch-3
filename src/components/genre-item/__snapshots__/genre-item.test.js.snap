// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreItem correctly renders after relaunch 1`] = `
<li
  className="catalog__genres-item"
>
  <a
    className="catalog__genres-link"
    onClick={[Function]}
  >
    
  </a>
</li>
`;
